File: app/src/commands/infrastructure/prefecture.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from '@tauri-apps/api/core';
import type {
  AddPrefectureBuildOrderRequest,
  CancelPrefectureBuildOrderRequest,
  Coord,
  GetPrefectureBuildCatalogRequest,
  GetPrefectureBuildCatalogResponse,
  PrefectureBuildOrderRequest,
} from '@tsukilabs/nil-bindings';

export async function addPrefectureBuildOrder(request: PrefectureBuildOrderRequest) {
  const req: AddPrefectureBuildOrderRequest = {
    world: NIL.world.getIdStrict(),
    request,
  };

  await invoke('add_prefecture_build_order', { req });
}

export async function cancelPrefectureBuildOrder(coord: Coord) {
  const req: CancelPrefectureBuildOrderRequest = {
    world: NIL.world.getIdStrict(),
    coord,
  };

  await invoke('cancel_prefecture_build_order', { req });
}

export async function getPrefectureBuildCatalog(coord: Coord) {
  const req: GetPrefectureBuildCatalogRequest = {
    world: NIL.world.getIdStrict(),
    coord,
  };

  return invoke<GetPrefectureBuildCatalogResponse>('get_prefecture_build_catalog', { req });
}

File: app/src-tauri/src/command/infrastructure/prefecture.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::manager::ManagerExt;
use nil_payload::request::infrastructure::prefecture::*;
use nil_payload::response::infrastructure::prefecture::*;
use tauri::AppHandle;

#[tauri::command]
pub async fn add_prefecture_build_order(
  app: AppHandle,
  req: AddPrefectureBuildOrderRequest,
) -> Result<()> {
  app
    .client(async |cl| cl.add_prefecture_build_order(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cancel_prefecture_build_order(
  app: AppHandle,
  req: CancelPrefectureBuildOrderRequest,
) -> Result<()> {
  app
    .client(async |cl| cl.cancel_prefecture_build_order(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_prefecture_build_catalog(
  app: AppHandle,
  req: GetPrefectureBuildCatalogRequest,
) -> Result<GetPrefectureBuildCatalogResponse> {
  app
    .client(async |cl| cl.get_prefecture_build_catalog(req).await)
    .await
    .map_err(Into::into)
}
